### ISSUE TITLE
SPID SP metadata generator - avviso SPID n. 29v3 enhancements

### ISSUE THREAD

**Issue by nicolabeghin** [lscorcia/keycloak-spid-provider#5]
Hi
first of all, kudos for such an excellent job in getting SPID provider for Keycloak. Really amazing work.
I would like to contribute to have the SP generator working as expected and as per latest requirements.

I'm currently extending the [SP metadata generator](https://github.com/nicolabeghin/keycloak-spid-provider/blob/spid_sp_metadata_enhancements/src/main/java/org/keycloak/broker/spid/metadata/SpidSpMetadataExtensionGenerator.java) in order to have it fully supported as per [avviso SPID n. 29v3](https://www.agid.gov.it/sites/default/files/repository_files/spid-avviso-n29v3-specifiche_sp_pubblici_e_privati.pdf)

WIP fork at https://github.com/nicolabeghin/keycloak-spid-provider/tree/spid_sp_metadata_enhancements

I have 2 questions:
1. do you know why the new `Contact` fields are not shown in Keycloak IdP configuration interface?  [they've been added](https://github.com/nicolabeghin/keycloak-spid-provider/commit/155e4dd2120d026626a4550f4cc74cb921576b74) by replicating the `Organization` fields, but still not showing up.
2. when trying to add the new `ContactType.Extensions` element, Keycloak is throwing an exception I'm not able to understand, not sure if it's linked to Java modules or Wildfly/Keycloak modules or silly classcast. [`SpidSpMetadataExtensionGenerator`](https://github.com/nicolabeghin/keycloak-spid-provider/blob/spid_sp_metadata_enhancements/src/main/java/org/keycloak/broker/spid/metadata/SpidSpMetadataExtensionGenerator.java) has been created using Keycloak's OOTB [`KeycloakKeySamlExtensionGenerator`](https://github.com/keycloak/keycloak/blob/master/saml-core/src/main/java/org/keycloak/saml/processing/core/util/KeycloakKeySamlExtensionGenerator.java)
![image](https://user-images.githubusercontent.com/2743637/107122449-203b3180-6898-11eb-8f0e-d598569ac7e8.png)

```
Caused by: java.lang.ClassCastException: class org.keycloak.broker.spid.metadata.SpidSpMetadataExtensionGenerator cannot be cast to class org.w3c.dom.Element (org.keycloak.broker.spid.metadata.SpidSpMetadataExtensionGenerator is in unnamed module of loader 'deployment.spid-provider.jar' @779584f; org.w3c.dom.Element is in module java.xml of loader 'bootstrap')
	at org.keycloak.keycloak-saml-core-public@12.0.2//org.keycloak.dom.saml.v2.metadata.ExtensionsType.getElement(ExtensionsType.java:55)
	at org.keycloak.keycloak-saml-core@12.0.2//org.keycloak.saml.processing.core.saml.v2.writers.SAMLMetadataWriter.write(SAMLMetadataWriter.java:439)
	at org.keycloak.keycloak-saml-core@12.0.2//org.keycloak.saml.processing.core.saml.v2.writers.SAMLMetadataWriter.writeEntityDescriptor(SAMLMetadataWriter.java:173)
	at deployment.spid-provider.jar//org.keycloak.broker.spid.metadata.SpidSpMetadataResourceProvider.getSPDescriptor(SpidSpMetadataResourceProvider.java:376)
	at deployment.spid-provider.jar//org.keycloak.broker.spid.metadata.SpidSpMetadataResourceProvider.get(SpidSpMetadataResourceProvider.java:185)
	... 74 more
```

thanks a lot, I think SP metadata is 95% done to have it honoring latest SPID requirements.

**Comment by nicolabeghin** [comment]
@lscorcia any input on this? I don't expect a solution of course, just a hint to get it going if you have any ideas! Thanks a lot

**Comment by lscorcia** [comment]
Sorry, I have been really busy with work at the moment and did not have enough time to check your progress yet. I am trying to find some spare time to work on this as I would really love to complete this feature. I hope to be able to take a look at it next weekend, but I can't make any promises! :)

**Comment by nicolabeghin** [comment]
Still unable to return this **** `Element`, always getting exception 
> `Caused by: javax.xml.stream.XMLStreamException: Trying to write an attribute when there is no open start element.`

Seems like the returned XML element is not properly handled/created leading to [mess up the whole `ContactType` rendering](https://github.com/keycloak/keycloak/blob/12.0.2/saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/writers/SAMLMetadataWriter.java#L434)!!

![image](https://user-images.githubusercontent.com/2743637/107687747-2ae63400-6ca7-11eb-8daf-37c510ed6555.png)

    package org.keycloak.broker.spid.metadata;
    
    import org.keycloak.dom.saml.v2.metadata.ExtensionsType;
    import org.w3c.dom.*;
    import javax.xml.parsers.DocumentBuilderFactory;
    
    public class SpidContactTypeExtension extends ExtensionsType {
    
        public static final String NS_URI = "https://spid.gov.it/saml-extensions";
        public static final String NS_PREFIX = "spid";
        private final String ELEMENT;
        private final String ELEMENT_VALUE;
    
        public SpidContactTypeExtension(String element, String value) {
            this.ELEMENT = element;
            this.ELEMENT_VALUE = value;
        }
    
        @Override
        public Element getElement() {
            try {
                DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
                builderFactory.setNamespaceAware(true);
                builderFactory.setValidating(false);
                Document document = builderFactory.newDocumentBuilder().newDocument();
                Element element = document.createElementNS(NS_URI, ELEMENT);
                element.setPrefix(NS_PREFIX);
                element.appendChild(document.createTextNode(ELEMENT_VALUE));
                return element;
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }

**Comment by lscorcia** [comment]
I have to say the whole handling of the metadata writer looks pretty strange to me. Keycloak has something that look like XJC-generated classes to represent the metadata schema, but then they use a custom hand-written writer instead of direct object serialization. Also, the way the current code is written, it does not seem possible to pass more than one single Extension (as you noticed, it uses the getElement method of the ExtensionsType which returns a single Element).

I will think about it and possibly prepare a patch to submit upstream. This means that even if it gets accepted we will need to wait for the full release of Keycloak 13 before adding this feature to the project. Bummer!

**Comment by nicolabeghin** [comment]
thanks @lscorcia, this reassures me I'm not completely dumb :D 

IMHO: the `getElement()` was something inherited from the past and never updated ([`ExtensionsType.getElement()`](https://github.com/keycloak/keycloak/blob/12.0.2/saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java#L54) is deprecated in favor of [`ExtentionsType.getAny()`](https://github.com/keycloak/keycloak/blob/12.0.2/saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java#L89)) therefore should provide a way to add more than one `Extension`. But as you outlined it would require a change in Keycloak itself (to avoid using `getElement()`) and understand how to properly add a correct Extension in the existing [`List`](https://github.com/keycloak/keycloak/blob/12.0.2/saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java#L47)

**Comment by lscorcia** [comment]
So, today I spent some time poking around Keycloak sources and there really are no good news. While I have been able to invoke JAXB serialization on an EntityDescriptor, which allows adding any required extension, it would be a pretty large patch to submit upstream. Before starting this work I will try to get some feedback on the Keycloak-dev mailing list.
I really have no idea why they wrote all of this code by hand. Maybe back then XJC was not mature enough :shrug:

**Comment by nicolabeghin** [comment]
FYI I committed a **very** quick and dirty workaround to get required Extensions in the generated SP metadata, plus the additional fields in provider configuration.

https://github.com/nicolabeghin/keycloak-spid-provider/commits/spid_sp_metadata_enhancements

**Comment by lscorcia** [comment]
I sent a PR upstream to allow metadata extensions with very little code: https://github.com/keycloak/keycloak/pull/7829 , but if merged it will be available in Keycloak 13.

I will try to use a similar technique to allow extensions in the current provider.

**Comment by lscorcia** [comment]
Fixed for Keycloak 12.x on the master branch, could you please give it a spin and let me know before I set up an official release and prepare the docs? I merged one of your commits BTW, so thanks a lot!

**Comment by nicolabeghin** [comment]
HUGE work, my most sincere compliments. I will test asap with the revised version.

**Comment by nicolabeghin** [comment]
Just moved to your updated branch in place of the silly/quick-and-dirty branch of mine, all good!

If useful to move over the IdP config to the new naming convention you adopted:

```
UPDATE IDENTITY_PROVIDER_CONFIG SET name='otherContactIpaCode' WHERE name='contactIPACode';
UPDATE IDENTITY_PROVIDER_CONFIG SET name='otherContactPhone' WHERE name='contactPhone';
UPDATE IDENTITY_PROVIDER_CONFIG SET name='otherContactEmail' WHERE name='contactEmail';
```

**Comment by lscorcia** [comment]
Released in version 1.0.7. Thanks for your help!
